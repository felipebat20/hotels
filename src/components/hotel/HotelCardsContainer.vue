<template>
  <div
    v-if="hotelsStore.control_flow.request_pending && ! hotelsStore.hotels.length"
    class="row justify-center q-my-md"
  >
    <q-spinner-dots color="primary" size="40px" />
  </div>

  <template v-else>
    <q-infinite-scroll
      @load="handleLoad"
      :offset="200"
      :disable="! hasMoreHotels"
    >
      <template v-if="hotelsStore.hotels.length">
        <HotelCard
          v-for="hotel in hotelsStore.hotels"
          :key="hotel.id"
          :hotel="hotel"
          @set-selected-hotel="handleHotelClick"
        />
      </template>


      <template v-else>
        <div class="column items-center q-pa-xl">
          <p class="text-h6 text-grey-6">
            Nenhum hotel encontrado
          </p>
        </div>
      </template>

      <template #loading>
        <div class="row justify-center q-my-md">
          <q-spinner-dots
            color="primary"
            size="40px"
          />
        </div>
      </template>
    </q-infinite-scroll>

    <div
      v-if="! hotelsStore.control_flow.has_next_page && hotelsStore.hotels.length"
      class="row justify-center"
    >
      <p class="text-grey-6">
        Não existem mais hotéis para serem exibidos
      </p>
    </div>
  </template>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';

import HotelCard from 'src/components/hotel/HotelCard.vue';

import { HotelInterface } from 'src/interfaces/HotelInterface';

import { useHotelsStore } from 'src/stores/useHotelsStore';

const hotelsStore = useHotelsStore();
const request_pending = ref(false);


const hasMoreHotels = computed(() => {
  return hotelsStore.control_flow.has_next_page;
});

const handleLoad = async (index: number, done: () => void) => {
  hotelsStore.control_flow.request_pending = 1 === index;
  request_pending.value = true;

  await hotelsStore.fetchHotels({
    sortBy: hotelsStore.control_flow.sortBy,
    name: hotelsStore.control_flow.name,
    placeId: hotelsStore.control_flow.placeId,
    page: index,
  });

  request_pending.value = false;

  done();
};

const emit = defineEmits(['set-selected-hotel']);
const handleHotelClick = (hotel: HotelInterface) => {
  emit('set-selected-hotel', hotel);
};
</script>
