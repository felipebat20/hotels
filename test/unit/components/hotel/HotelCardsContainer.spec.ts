import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest';
import { mount } from '@vue/test-utils';
import { describe, expect, it, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing'

import HotelCardsContainer from '@/components/hotel/HotelCardsContainer.vue';

installQuasarPlugin();

describe('test HotelCardsContainer', () => {
  const createWrapper = () => {
    return mount(HotelCardsContainer, {
      global: {
        plugins: [createTestingPinia({ createSpy: vi.fn })],
      },
    });
  };

  it('should mount component properly', () => {
    const wrapper = createWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
