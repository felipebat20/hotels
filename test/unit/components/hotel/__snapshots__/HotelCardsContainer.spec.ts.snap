// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test HotelCardsContainer > should mount component properly 1`] = `
"<div class="q-infinite-scroll">
  <div class="column items-center q-pa-xl">
    <p class="text-h6 text-grey-6"> Nenhum hotel encontrado </p>
  </div>
  <div class="q-infinite-scroll__loading invisible">
    <div class="row justify-center q-my-md"><svg class="q-spinner text-primary" fill="currentColor" width="40px" height="40px" viewBox="0 0 120 30" xmlns="http://www.w3.org/2000/svg">
        <circle cx="15" cy="15" r="15">
          <animate attributeName="r" from="15" to="15" begin="0s" dur="0.8s" values="15;9;15" calcMode="linear" repeatCount="indefinite"></animate>
          <animate attributeName="fill-opacity" from="1" to="1" begin="0s" dur="0.8s" values="1;.5;1" calcMode="linear" repeatCount="indefinite"></animate>
        </circle>
        <circle cx="60" cy="15" r="9" fill-opacity=".3">
          <animate attributeName="r" from="9" to="9" begin="0s" dur="0.8s" values="9;15;9" calcMode="linear" repeatCount="indefinite"></animate>
          <animate attributeName="fill-opacity" from=".5" to=".5" begin="0s" dur="0.8s" values=".5;1;.5" calcMode="linear" repeatCount="indefinite"></animate>
        </circle>
        <circle cx="105" cy="15" r="15">
          <animate attributeName="r" from="15" to="15" begin="0s" dur="0.8s" values="15;9;15" calcMode="linear" repeatCount="indefinite"></animate>
          <animate attributeName="fill-opacity" from="1" to="1" begin="0s" dur="0.8s" values="1;.5;1" calcMode="linear" repeatCount="indefinite"></animate>
        </circle>
      </svg></div>
  </div>
</div>
<!--v-if-->"
`;
